src/plexil-adapter/OwAdapter.cpp

```diff
@@ -78,19 +85,35 @@ static void ack_sent (Command* cmd, AdapterExecInterface* intf)
   ack_command (cmd, COMMAND_SENT_TO_SYSTEM, intf);
 }
 
-static void command_status_callback (int id, bool success)
+static void send_ack_once(CommandRecord& cr, bool skip=false)
 {
-  pair<Command*, AdapterExecInterface*> cmd;
-
-  try {
-    cmd = CommandRegistry.at(id);
+  std::lock_guard<std::mutex> g(g_shared_mutex);
+  bool& sent_flag = std::get<CR_ACK_SENT>(cr);
+  if (!sent_flag)
+  {
+    if (!skip) {
+      ack_sent(std::get<CR_COMMAND>(cr), std::get<CR_ADAPTER>(cr));
+    }
+    sent_flag = true;
   }
-  catch (const std::out_of_range& oor) {
-    ROS_ERROR("command_status_callback: no command registered under id %d", id);
+}
+
+static void command_status_callback (int id, bool success)
+{
+  auto it = CommandRegistry.find(id);
+  if (it == CommandRegistry.end())
+  {
+    ROS_ERROR_STREAM("command_status_callback: no command registered under id"
+                     << id);
     return;
   }
-  if (success) ack_success (cmd.first, cmd.second);
-  else ack_failure (cmd.first, cmd.second);
+
+  std::unique_ptr<CommandRecord>& cr = it->second;
+  Command* cmd = std::get<CR_COMMAND>(*cr);
+  AdapterExecInterface* intf = std::get<CR_ADAPTER>(*cr);
+  send_ack_once(*cr, true);
```

Comment:
I'm pretty sure that send_ack_once is never needed in this function, but it is not hurting anything -- just causing an extra executive cycle.

Reply:
The only case where I am able to produce the stalling issue is when the two signal come out of order.